dds/DCPS/DiscoveryBase.h

```diff
@@ -1263,6 +1201,263 @@ namespace OpenDDS {
+                  reader_participant, dwch, relay_only_readers_.count(reader));
+              remote_reader_crypto_handles_[reader] = drch;
+              DatareaderCryptoTokenSeqMap::iterator t_iter =
+                pending_remote_reader_crypto_tokens_.find(reader);
+              if (t_iter != pending_remote_reader_crypto_tokens_.end()) {
+                DDS::Security::SecurityException se;
+                if (!keyexg->set_remote_datareader_crypto_tokens(iter->second, drch, t_iter->second, se)) {
+                  ACE_ERROR((LM_ERROR, ACE_TEXT("(%P|%t) ERROR: ")
+                    ACE_TEXT("(%P|%t) ERROR: DiscoveryBase::match() - ")
+                    ACE_TEXT("Unable to set pending remote datareader crypto tokens with crypto ")
+                    ACE_TEXT("key exchange plugin. Security Exception[%d.%d]: %C\n"),
+                    se.code, se.minor_code, se.message.in()));
+                }
+                pending_remote_reader_crypto_tokens_.erase(t_iter);
+              }
+              EndpointSecurityAttributesMap::const_iterator s_iter =
+                local_writer_security_attribs_.find(writer);
+              if (s_iter != local_writer_security_attribs_.end() &&
+                (s_iter->second.is_submessage_protected || s_iter->second.is_payload_protected)) {
+                create_and_send_datawriter_crypto_tokens(dwch, writer, drch, reader);
+              }
+            }
+          }
+        }
+#endif
+
+        // Copy reader and writer association data prior to releasing lock
+#ifdef __SUNPRO_CC
+        ReaderAssociation ra;
+        ra.readerTransInfo = *rTls;
+        ra.readerId = reader;
+        ra.subQos = *subQos;
+        ra.readerQos = *drQos;
+        ra.filterClassName = cfProp->filterClassName;
+        ra.filterExpression = cfProp->filterExpression;
+        ra.exprParams = cfProp->expressionParameters;
+        XTypes::serialize_type_info(reader_type_info, ra.serializedTypeInfo);
+        WriterAssociation wa;
+        wa.writerTransInfo = *wTls;
+        wa.writerId = writer;
+        wa.pubQos = *pubQos;
+        wa.writerQos = *dwQos;
+        XTypes::serialize_type_info(writer_type_info, wa.serializedTypeInfo);
+#else
+        DDS::OctetSeq octet_seq_type_info_reader;
+        XTypes::serialize_type_info(*reader_type_info, octet_seq_type_info_reader);
+        const ReaderAssociation ra =
+        { add_security_info(*rTls, writer, reader), reader, *subQos, *drQos,
+#ifndef OPENDDS_NO_CONTENT_FILTERED_TOPIC
+             cfProp->filterClassName, cfProp->filterExpression,
+#else
+             "", "",
+#endif
+             cfProp->expressionParameters, octet_seq_type_info_reader };
+
+        DDS::OctetSeq octet_seq_type_info_writer;
+        XTypes::serialize_type_info(*writer_type_info, octet_seq_type_info_writer);
+        const WriterAssociation wa =
+        { add_security_info(*wTls, writer, reader), writer, *pubQos, *dwQos, octet_seq_type_info_writer };
+#endif
```

Comment:
Continuation of [this](https://github.com/objectcomputing/OpenDDS/pull/1840/files#r470381737)

Here's the change, also with some formatting tweaks to the non-sun code since it's off:

```suggestion
        DDS::OctetSeq octet_seq_type_info_reader;
        XTypes::serialize_type_info(*reader_type_info, octet_seq_type_info_reader);
        const ReaderAssociation ra = {
          add_security_info(*rTls, writer, reader), reader, *subQos, *drQos,
#ifndef OPENDDS_NO_CONTENT_FILTERED_TOPIC
          cfProp->filterClassName, cfProp->filterExpression,
#else
          "", "",
#endif
          cfProp->expressionParameters,
          octet_seq_type_info_reader
        };

        DDS::OctetSeq octet_seq_type_info_writer;
        XTypes::serialize_type_info(*writer_type_info, octet_seq_type_info_writer);
        const WriterAssociation wa = {
          add_security_info(*wTls, writer, reader), writer, *pubQos, *dwQos,
          octet_seq_type_info_writer
        };
```